src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+
+pub const DEFAULT_BUF_SIZE: usize = 64 * 1024;
+
+
+pub fn read_to_end_zeroed<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> io::Result<usize> {
```

Comment:
Unfortunately making read_to_end take a trait object doesn't currently appear to be possible: if I try I get the following error:

src/libstd/sys/common/io.rs:19:26: 19:27 error: the trait `core::marker::Sized` is not implemented for the type `R` [E0277]
src/libstd/sys/common/io.rs:19     unsafe { read_to_end(r, buf, true) }

Asked on IRC with a reduced example ( https://play.rust-lang.org/?code=use%20std%3A%3Aio%3A%3ARead%3B%0A%0Afn%20foo%28r%3A%20%26mut%20Read%29%20{%0A%0A}%0A%0Afn%20bar%3CR%3A%20Read%2B%3FSized%3E%28r%3A%20%26mut%20R%29%20{%0A%20%20%20%20foo%28r%29%3B%0A}%0A%0Afn%20main%28%29%20{%0A%20%20%20%20bar%28%26mut%20std%3A%3Aio%3A%3Astdin%28%29%29%3B%0A}&version=stable ) and nobody seemed quite sure why that wouldn't work, given that R is behind a reference.